tensorflow-framework/src/main/java/org/tensorflow/framework/callbacks/util/ProgressBar.java

```diff
@@ -0,0 +1,489 @@
+  private final List<String> valuesOrdered = new ArrayList<>();
+  private final long start = System.currentTimeMillis();
+  // will be null if java system console is not present
+  private final Console console;
+  // defaults to stdout if console is null;
+  private final PrintWriter writer;
+  private final boolean dynamicDisplay;
+  private Integer target;
+  private int totalWidth;
+  private int seenSoFar;
+  private long lastUpdate;
+  private Long timeAfterFirstStep;
+  /** Create a ProgressBar */
+  public ProgressBar() {
+    this(null, DEFAULT_WIDTH, DEFAULT_VERBOSE, DEFAULT_INTERVAL, DEFAULT_COUNT_MODE, null);
+  }
+  /**
+   * Create a ProgressBar
+   *
+   * @param target Total number of steps expected, null if unknown.
+   */
+  public ProgressBar(Integer target) {
+    this(target, DEFAULT_WIDTH, DEFAULT_VERBOSE, DEFAULT_INTERVAL, DEFAULT_COUNT_MODE, null);
+  }
+
+  /**
+   * Create a ProgressBar
+   *
+   * @param target Total number of steps expected, null if unknown.
+   * @param verbose Verbosity mode
+   * @param unit Display name for step counts, "step" or "sample".
+   * @param statefulMetrics names of metrics that should not be averaged over an epoch. Metrics in
+   *     this list will be logged as-is. All others will be averaged over time (e.g. loss, etc). If
+   *     not provided, defaults to the Model's metrics.
+   */
+  public ProgressBar(
+      Integer target, VerboseMode verbose, CountMode unit, Set<String> statefulMetrics) {
+    this(target, DEFAULT_WIDTH, verbose, DEFAULT_INTERVAL, unit, statefulMetrics);
+  }
+
+  /**
+   * Create a ProgressBar
+   *
+   * @param target Total number of steps expected, null if unknown.
+   * @param width Progress bar width on screen.
+   * @param verbose Verbosity mode, false is silent
+   * @param interval Minimum visual progress update interval (in milliseconds).
+   * @param unit Display name for step counts, "step" or "sample".
+   * @param statefulMetrics names of metrics that should not be averaged over an epoch. Metrics in
+   *     this list will be logged as-is. All others will be averaged over time (e.g. loss, etc). If
+   *     not provided, defaults to the Model's metrics.
+   */
+  public ProgressBar(
+      Integer target,
+      int width,
+      VerboseMode verbose,
+      double interval,
+      CountMode unit,
+      Set<String> statefulMetrics) {
+    this.target = target;
+    this.width = width;
+    this.verbose = verbose;
+    this.interval = (long) interval;
+    this.unit = unit;
+    this.statefulMetrics = statefulMetrics == null ? Collections.emptySet() : statefulMetrics;
+    switch (verbose) {
+      case VERBOSE:
+        console = System.console();
+        writer = console != null ? null : new PrintWriter(System.out);
+        break;
+      case SEMI_VERBOSE:
+        writer = new PrintWriter(System.out);
+        console = null;
+        break;
+      default:
+        writer = null;
```

Comment:
Won't this default cause an NPE in `print`? Maybe it should throw `IllegalStateException`? That way it blows up quickly at least.